src/clipping/greinerhormann.jl

```diff
@@ -0,0 +1,883 @@
+    next,
+    neighbour,
+    source,
+    intersection,
+    alpha,
+    label,
+    enex
+  )
+
+  GreinerVertex() = new(
+    Point((0, 0)),
+    1,
+    1,
+    1,
+    false,
+    false,
+    -1.0,
+    IntersectionLabel(0),
+    EntryExit(2)
+  )
+end
+
+function GreinerVertex(coords::Point)
+  GreinerVertex(coords, 1, 1, 1, false, false, -1.0, IntersectionLabel(0), EntryExit(2))
+end
+
+
+
+function GreinerVertex(ring::T) where {T<:Ring}
+  verts = vertices(ring)
+  ringₜ = Vector{GreinerVertex}()
+  ringₜ = _newVertex!(ringₜ, verts[1], true)
+  for i in 2:length(verts)
+    prev_index = i == 1 ? length(verts) : i - 1
+    next_index = i == length(verts) ? 1 : i + 1
+    _newVertex!(ringₜ, verts[i], false)
+  end
+  ringₜ
+
+end
+
+_getneighbor(v::GreinerVertex) = v.neighbour
+_getprev(v::GreinerVertex) = v.prev
+_getnext(v::GreinerVertex) = v.next
+coords(v::GreinerVertex) = v.coords
+_getenex(v::GreinerVertex) = v.enex
+
+_setneighbor(v::GreinerVertex, n::Int) = v.neighbour = n
+_setprev(v::GreinerVertex, p::Int) = v.prev = p
+_setnext(v::GreinerVertex, n::Int) = v.next = n
+_setcoords(v::GreinerVertex, c::Point) = v.coords = c
+_setenex(v::GreinerVertex, e::EntryExit) = v.enex = e
+
+Line(p₁::GreinerVertex, p₂::GreinerVertex) = Line(coords(p₁), coords(p₂))
+vertices(ring::Vector{GreinerVertex}) = [coords(v) for v in ring]
+
+
+
+#inserts p₁ after current
+function _insertVertex!(p₁::GreinerVertex, current::GreinerVertex, vertex_vector::Vector{GreinerVertex})
+  verticesₜ = vertices(vertex_vector)
+  idx = current.next
+  p₁.prev = idx - 1
+  p₁.next = idx + 1
+  # insert p₁ at idx
+  insert!(vertex_vector, idx, p₁)
+  # update vertex_vector, any vertex next should point one index ahead
+  for i in (idx+1):length(vertex_vector)
+    vertex_vector[i].next += 1
+    vertex_vector[i].prev += 1
+
+  end
+end
+
+
+function _link(v₁::GreinerVertex, v₂::GreinerVertex, verticesS::Vector{GreinerVertex}, verticesC::Vector{GreinerVertex})
+  v₁.neighbour = findfirst(x -> x == coords(v₂), vertices(verticesC))
+  v₂.neighbour = findfirst(x -> x == coords(v₁), vertices(verticesS))
+  v₁.intersection = true
+  v₂.intersection = true
+end
+
+function _newVertex!(ring::Vector{GreinerVertex}, coords::Point, source=false)
+  vertex = GreinerVertex()
+  vertex.coords = coords
+  vertex.source = source
+  if isempty(ring)
+    vertex.prev = 1
+    vertex.next = 1
+    push!(ring, vertex)
+  else
+    # add vertex at the end
+    vertex.prev = length(ring)
+    vertex.next = 1
+    ring[end].next = length(ring) + 1
+    ring[1].prev = length(ring) + 1
+    push!(ring, vertex)
+  end
+end
+
+function _removeVertex(ring::Vector{GreinerVertex}, index::Int)
+  if length(ring) == 1
+    return Vector{GreinerVertex}()
+  end
+
+  prev_index = ring[index].prev
+  next_index = ring[index].next
+
+  ring[prev_index].next = next_index
+  ring[next_index].prev = prev_index
+
+  deleteat!(ring, index)
+  return ring
+end
+
+function pointInPoly(R::Point, poly::Ring)
```

Comment:
here you mean `point \in poly` maybe? we already have a super efficient (multi-threaded) implementation for this